### ISSUE TITLE
IPv6 on GitHub-hosted runners

### ISSUE THREAD

**Issue by BrightRan** [actions/virtual-environments#668]
This issue is transferred from [here](https://github.com/actions/runner/issues/402).

In [this ticket](https://github.community/t5/GitHub-Actions/GitHub-Actions-IPv6-Network-Issues/td-p/52085), the user is trying to reach the **`IPv6 address`** of Google by running the command "**`curl -6 https://www.google.com`**" on the **`GitHub-hosted Ubuntu runner`**, but it always failed to connect and returned the message "**`Couldn't connect to server`**".
The command "**`curl https://www.google.com`**" or "**`curl -4 https://www.google.com`**" can work fine to reach the **`IPv4 address`**.

I also tested directly ping the **`IPv6 address`** of Google via the command "**`ping6 ipv6.google.com`**", it also failed to connect and returned the message "**`connect: Network is unreachable`**".

To check whether the current Linux kernel supports **`IPv6`** via the command,
```
[ -f /proc/net/if_inet6 ] && echo 'IPv6 ready system!' || echo 'No IPv6 support found! Compile the kernel!!'
```
It returns "**`IPv6 ready system!`**", looks like **`IPv6`** is supported.
Then to check whether the **`IPv6 module`** has loaded,
```
lsmod | grep -qw ipv6 && echo "IPv6 kernel driver loaded and configured." || echo "IPv6 not configured and/or driver loaded on the system."
```
It returns "**`IPv6 not configured and/or driver loaded on the system.`**", looks like the **`IPv6 module`** is not loaded.
Load the **`IPv6 module`** and check again.
```
modprobe ipv6
lsmod | grep -qw ipv6 && echo "IPv6 kernel driver loaded and configured." || echo "IPv6 not configured and/or driver loaded on the system."
```
The returned result still shows that the **`IPv6 module`** is not loaded. Looks like this should be the reason for failed to reach the **`IPv6 address`**.

Why it can't load the **`IPv6 module`** on the GitHub-hosted runner? Whether this is an expected behavior for some limitations set on the hosted runners?

**Comment by miketimofeev** [comment]
@BrightRan looks like the images itself are ok. 
@alepauly do we have ipv6 configured on the environment side?

**Comment by airtower-luna** [comment]
[According to the Docker documentation](https://docs.docker.com/config/daemon/ipv6/) the images would still need to enable IPv6 for Docker containers, even if the VMs start to support it.

**Comment by alepauly** [comment]
Unfortunately we can't support this at this time due to infrastructure constraints.

**Comment by lzhecheng** [comment]
> Unfortunately we can't support this at this time due to infrastructure constraints.

Is it in the roadmap? I found no opening issues on IPv6 support. Thanks.

**Comment by miyurusankalpa** [comment]
GitHub IP address list now has IPv6 address for GitHub Actions.
https://api.github.com/meta

**Comment by oscartbeaumont** [comment]
As of a week ago, I could not get IPv6 working in GitHub Actions but let's hope them listing those address's means they are about to release support for it.

**Comment by pavelkim** [comment]
Tested today.
Under both ubuntu-latest and ubuntu-20.04:

`curl -vv --ipv6 "[2607:f8b0:4002:c03::93]:80"`
```
* Rebuilt URL to: [2607:f8b0:4002:c03::93]:80/
*   Trying 2607:f8b0:4002:c03::93...
* TCP_NODELAY set
* Immediate connect fail for 2607:f8b0:4002:c03::93: Network is unreachable
* Closing connection 0
curl: (7) Couldn't connect to server
```

`ip route sh`
```
default via 10.1.0.1 dev eth0 proto dhcp src 10.1.0.4 metric 100 
10.1.0.0/16 dev eth0 proto kernel scope link src 10.1.0.4 
168.63.129.16 via 10.1.0.1 dev eth0 proto dhcp src 10.1.0.4 metric 100 
169.254.169.254 via 10.1.0.1 dev eth0 proto dhcp src 10.1.0.4 metric 100 
172.17.0.0/16 dev docker0 proto kernel scope link src 172.17.0.1 linkdown 
```

`ip addr sh`
```
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host 
       valid_lft forever preferred_lft forever
2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq state UP group default qlen 1000
    link/ether 00:0d:3a:c6:7b:12 brd ff:ff:ff:ff:ff:ff
    inet 10.1.0.4/16 brd 10.1.255.255 scope global eth0
       valid_lft forever preferred_lft forever
    inet6 fe80::20d:3aff:fec6:7b12/64 scope link 
       valid_lft forever preferred_lft forever
3: docker0: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc noqueue state DOWN group default 
    link/ether 02:42:79:06:ee:ce brd ff:ff:ff:ff:ff:ff
    inet 172.17.0.1/16 brd 172.17.255.255 scope global docker0
       valid_lft forever preferred_lft forever
```

No IPv6 routes so far.

**Comment by vistart** [comment]
Is there any way to get only the IPv4 address of the domain name at the VM level?

**Comment by NiKiZe** [comment]
@alepauly Time to reopen this! it is 2022 IPv6 is a requirement.
Legacy-v4 might live for a long time (to long) but you can't ignore v6!

**Comment by pcgeek86** [comment]
Almost the end of January 2022 ... still no IPv6?

**Comment by idealo-ricardo-band** [comment]
I just found out about the non-support for the current Internet Protocol that is now nearly 25 years old.
Github, this is really embarrassing. There really is no excuse to not support that for years. Please get your shit together. Pardon my french..

**Comment by XenGi** [comment]
I just found out about the non-support for the current Internet Protocol that is now nearly 25 years old.
Github, this is really embarrassing. There really is no excuse to not support that for years. Please get your shit together. Pardon my french..

**Comment by pradt2** [comment]
Summer 2022 and still no IPv6 support. Unacceptable.

**Comment by XenGi** [comment]
@alepauly I don't know why you closed this ticket in the first place but the demand for IPv6 is just getting higher. Do you see any chance of this ever getting fixed or should we all just switch to Gitlab CI or go back to Travis CI.

**Comment by idealo-ricardo-band** [comment]
Any updates on this?

**Comment by NiKiZe** [comment]
GitHub and the won't fix from @alepauly shows that they have no intention to fixing this. If they did this issue would have remained open.

**Comment by idealo-ricardo-band** [comment]
> Unfortunately we can't support this at this time due to infrastructure constraints.

So Github is not able to adapt it's infrastructure in over 2 years. Wow..

**Comment by Otus9051** [comment]
Its just IPv6, its not even like they need to change their BGP peers or something.
Same with Azure, they still don't provide IPv6 by default with their services.

**Comment by NiKiZe** [comment]
Maybe, just maybe there is a little bit of hope on this moving forward. https://github.com/orgs/community/discussions/10539#discussioncomment-3917345

**Comment by ldemailly** [comment]
2023 ping - any chance to get runners able to do ipv6? I need it for testing v6 code

For others who might know a workaround: is there a v6 test env easily accessible/compatible with github action syntax?

**Comment by XenGi** [comment]
> For others who might know a workaround: is there a v6 test env easily accessible/compatible with github action syntax?

There is always Gitlab. It's better anyway imho.

**Comment by JeanCarlosChavarriaHughes** [comment]
> > For others who might know a workaround: is there a v6 test env easily accessible/compatible with github action syntax?
> 
> There is always Gitlab. It's better anyway imho. They sadly also don't have IPv6 in their hosted runners but at least you can set it up in your own runners.


I ended up moving some projects to GitLab and now we are able to perform the required testing with IPv6.

**Comment by tbraun96** [comment]
Sad!

**Comment by w8fyz** [comment]
Please GitHub 🙏

**Comment by jcamiel** [comment]
Hi guys, I was trying to add integration test on [Hurl](https://github.com/Orange-OpenSource/hurl) a CLI HTTP client based on libcurl.

What does surprise me is that I've been able to make a GET request using curl and ipv6 from a GitHub-hosted macOS runner:

curl on macOS is available under:

- /usr/bin/curl 8.1.2
- /usr/local/opt/curl/bin/curl 8.3.0

Both binary have IPv6 feature enabled but only curl 8.3.0 is able to GET https://google.com with `--ipv6` on GitHub macOS runners.

```shell
$ /usr/local/opt/curl/bin/curl --version
curl 8.3.0 (x86_64-apple-darwin21.6.0) libcurl/8.3.0 (SecureTransport) OpenSSL/3.1.2 zlib/1.2.11 brotli/1.1.0 zstd/1.5.5 libidn2/2.3.4 libssh2/1.11.0 nghttp2/1.56.0 librtmp/2.3 OpenLDAP/2.6.6
Release-Date: 2023-09-13
Protocols: dict file ftp ftps gopher gophers http https imap imaps ldap ldaps mqtt pop3 pop3s rtmp rtsp scp sftp smb smbs smtp smtps telnet tftp
Features: alt-svc AsynchDNS brotli GSS-API HSTS HTTP2 HTTPS-proxy IDN IPv6 Kerberos Largefile libz MultiSSL NTLM NTLM_WB SPNEGO SSL threadsafe TLS-SRP UnixSockets zstd
$ /usr/bin/curl --version
curl 8.1.2 (x86_64-apple-darwin21.0) libcurl/8.1.2 (SecureTransport) LibreSSL/3.3.6 zlib/1.2.11 nghttp2/1.45.1
Release-Date: 2023-05-30
Protocols: dict file ftp ftps gopher gophers http https imap imaps ldap ldaps mqtt pop3 pop3s rtsp smb smbs smtp smtps telnet tftp
Features: alt-svc AsynchDNS GSS-API HSTS HTTP2 HTTPS-proxy IPv6 Kerberos Largefile libz MultiSSL NTLM NTLM_WB SPNEGO SSL threadsafe UnixSockets
```

And test with `--ipv6` against google.com:

```shell
$/usr/local/opt/curl/bin/curl --ipv6 --verbose https://google.com/
...
Connected to google.com (::ffff:142.250.115.101) port 443
...
```

```shell
$/usr/bin/curl --ipv6 --verbose https://google.com/
curl: (7) Couldn't connect to server
```

So, it seems there is _kind_ of IPv6 support inside GitHub infra. I'm surprised to see that between curl 8.1.2 and curl 8.3.0 there is a difference of usage with `--ipv6` option but I can't explain it.

**Comment by cross** [comment]
You'll note however, that's not really IPv6.  It's an IPv4-mapped address.  If it were really IPv6, it would look like this, from my desktop Macbook:
```
$ curl -V
curl 8.1.2 (x86_64-apple-darwin23.0) libcurl/8.1.2 (SecureTransport) LibreSSL/3.3.6 zlib/1.2.12 nghttp2/1.55.1
[...]
$ curl --ipv6 --verbose https://google.com/
*   Trying [2607:f8b0:4002:c2c::65]:443...
* Connected to google.com (2607:f8b0:4002:c2c::65) port 443 (#0)
[...]
```

**Comment by i18n-now** [comment]
<img width="1057" alt="image" src="https://github.com/actions/runner-images/assets/145643935/bca83802-2f4b-4275-95c1-8a16a29aa071">
use this  

```
  jobs:
    ipv6:
      runs-on: ubuntu-latest
      steps:
        - name: Setup WARP
          uses: fscarmen/warp-on-actions@v2
        - name: ipv6
          run: |
            curl -m 9 --ipv6 --verbose https://google.com
            curl -m 9 -6 https://ifconfig.co
```

**Comment by undergroundwires** [comment]
Here's my workaround (open-source and documented) that I hope that can help you too:

- [privacy.sexy | force-ipv4](https://github.com/undergroundwires/privacy.sexy/tree/master/.github/actions/force-ipv4): GitHub action to use for GitHub workflows.
- [privacy.sexy | `force-ipv4.sh`](https://github.com/undergroundwires/privacy.sexy/blob/master/.github/actions/force-ipv4/force-ipv4.sh).: Script for general use.

After days of research and trial/error, this is how I got this working:

1. Create a script called [`force-ipv4.sh`](https://github.com/undergroundwires/privacy.sexy/tree/master/.github/actions/force-ipv4), that configures system to prefer IPv4 over IPv6, call it to configure the machine. It was not easy to find a reliable cross-platform solution and I went Cloudflare WARP for DNS resolution along with some system configurations.
2. To easily use the script in GitHub workflows, create GitHub action called [`force-ipv4`](https://github.com/undergroundwires/privacy.sexy/blob/master/.github/actions/force-ipv4/force-ipv4.sh) and call it in GitHub runners.
3. Fixes the IPv6 request issues, and you can happily run e.g. `fetch` API from Node.

Related commit introducing this fix: https://github.com/undergroundwires/privacy.sexy/commit/52fadcd6177ed06216be9c67dad57192ae02a4f9

**Comment by NiKiZe** [comment]
I'm sure that GitHub are using current internet standards by now right?
If not they might as well disable TLS1.2 as well and only allow TLS1.0 max, I'm sure that would be acceptable by everyone right?

**Comment by WitchAurora** [comment]
> I'm sure that GitHub are using current internet standards by now right? If not they might as well disable TLS1.2 as well and only allow TLS1.0 max, I'm sure that would be acceptable by everyone right?

This issue is about IPv6 on GitHub-hosted runners.

**Comment by WitchAurora** [comment]
> If not they might as well disable TLS1.2 as well and only allow TLS1.0 max, I'm sure that would be acceptable by everyone right?

It unlikely they disable TLS 1.2

**Comment by NiKiZe** [comment]
Maybe a timeline helps...
December 1998 RFC2460 IPv6
January 1999 RFC2246 TLS 1.0
April 2006 RFC4346 TLS 1.1
August 2008 RFC5246 TLS 1.2
July 2017 RFC8200 IPv6 (Next gen removed, effectively marking IPv4 as obsolete)
August 2018 RFC8446 TLS 1.3

Let me rephrase, Why would anyone accept non support for TLS1.2 or even TLS1.3 on any runners? Of course it is unacceptable an therefore not an issue.
Then why would non working IPv6 be acceptable?

Non IPv6 support today is even worse than non TLS1.2 support!

**Comment by WitchAurora** [comment]
> Let me rephrase, Why would anyone accept non support for TLS1.2 or even TLS1.3 on any runners? Of course it is unacceptable an therefore not an issue. Then why would non working IPv6 be acceptable?
> 
> Non IPv6 support today is even worse than non TLS1.2 support!

In 2020 they said it because of infrastructure https://github.com/actions/runner-images/issues/668#issuecomment-624080758
